### DISCUSSION TITLE
Ssh behaving weird from Mac to linux.

### DISCUSSION THREAD

**Discussion by limonoct** [ghostty-org/ghostty#3911] [Q&A] [answered]
Hi,

After following the steps to set up SSH (infocmp -x | ssh YOUR-SERVER -- tic -x -), I encountered some peculiar issues when connecting to my Linux host from my Mac. While the terminal displayed the correct colors, the behavior was strange. For instance, pressing the backspace key in the terminal would result in a space instead of a backspace. Additionally, when executing commands like “ls,” it would print “lsls” instead of the expected “ls.”

Furthermore, the prompt format (p10k) appeared to be incorrect, lacking color support.

Here’s an example of the issue when I executed “ls” (notice how it displays lsls, and the prompt arrow do not have color):
![Screenshot 2024-12-29 at 9 27 37 AM](https://github.com/user-attachments/assets/59d7ab0a-f213-43bf-bc4d-a0fde720f98e)

To resolve these issues, I implemented a workaround by adding the “xterm-256color” option to my SSH configuration. This modification resulted in normal behavior, and the prompt functioned as expected. However, the colors were slightly different.

Here’s an example of the prompt with the “xterm-256color” option (notice the blue colors are different):
![Screenshot 2024-12-29 at 9 29 26 AM](https://github.com/user-attachments/assets/dc4552a7-d193-4ad3-bafe-0769c5b8e74f)

For reference, here’s how the prompt should look (this is on my mac):
![Screenshot 2024-12-29 at 9 31 29 AM](https://github.com/user-attachments/assets/0a674e12-4102-472f-824d-e889ec21fcd5)

any workaround? 

Thanks!

**Comment by TApplencourt** [comment]
Oh, I have exactly the same "delete as space" as you! 

The `export TERM=xterm-256color` fixed the issue for me. Thanks!

**Reply by Lira-tron** [reply]
For me too, just the colors now are different

**Reply by alecdwm** [reply]
I added `term=xterm-256color` to my `$HOME/Library/Application\ Support/com.mitchellh.ghostty/config` file as an alternative, which also worked!

**Comment by feekApp** [comment]
Hi,

thats because on your server does not recognize xterm-ghostty,
see: https://ghostty.org/docs/help/terminfo#ssh

use as fallback in your .ssh-config as last lines:
```
# .ssh/config
Host *
  SetEnv TERM=xterm-256color
```

**Reply by Lira-tron** [reply]
Yes I added that option also but it messes up the colors, there is another one infocmp -x | ssh YOUR-SERVER -- tic -x -, which fixes the colors but is not working correclty

**Reply by feekApp** [reply]
when using `infocmp -x | ssh YOUR-SERVER -- tic -x -` you do not need the settings in your `.ssh/config`, because that will force your `$TERM` to `xterm-256color`!

**Reply by edrozenberg** [reply]
If the `infocmp` is not working as expected, double-check to make sure it created the necessary files in the terminfo directory,  wherever it is on your target Linux system (normally `/usr/share/terminfo/` or possibly `/usr/local/share/terminfo/`.

On my Linux systems I now have the following (an alias in `g` and the actual terminfo file in `x`):
```
# ls -al /usr/share/terminfo/g/ghost*
lrwxrwxrwx 1 root root 18 Dec 27 16:06 /usr/share/terminfo/g/ghostty -> ../x/xterm-ghostty

# ls -al /usr/share/terminfo/x/xterm-ghostty 
-rw-r--r-- 1 root root 3.8K Dec 27 15:48 /usr/share/terminfo/x/xterm-ghostty
```

**Reply by limonoct** [reply]
Thank you! 
they are not there, but there is no error.. how to recreate them?

**Reply by edrozenberg** [reply]
No error is weird. Make sure to ssh connect as root when sending them over:
`infocmp -x | ssh root@myservername -- tic -x -`
(and do this from a Ghostty terminal window, and make sure you didn't set something like `term=xterm-256color` in your Ghostty `config`)
Otherwise without root the target system will not be able to put new items in terminfo.

https://ghostty.org/docs/help/terminfo

In this Ghostty terminfo doc page there's a paragraph dealing with Mac systems older than Sonoma, read the info there, it starts with this sentence:
> macOS versions before Sonoma cannot use the system-bundled infocmp.

**Reply by limonoct** [reply]
actually, I am getting this error: "<stdin>", line 2, col 31, terminal 'xterm-ghostty': older tic versions may treat the description field as an alias

**Reply by edrozenberg** [reply]
Sounds like that issue with an older MacOS version. From the Ghostty terminfo docs:
> macOS versions before Sonoma cannot use the system-bundled infocmp. The bundled version of ncurses is too old to emit a terminfo entry that can be read by more recent versions of tic, and the command will fail with a bunch of Illegal character messages. You can fix this by using Homebrew to install a recent version of ncurses and replacing infocmp above with the full path /opt/homebrew/opt/ncurses/bin/infocmp.

This is is for Apple Silicon. For Intel Macs you'd need to use `/usr/local/bin/infocmp` instead, after you install ncurses with homebrew.

**Reply by limonoct** [reply]
so, in the linux machine it is actually getting installed 

 infocmp ghostty
#       Reconstructed via infocmp from file: /home/linuxbrew/.linuxbrew/Cellar/ncurses/6.5/share/terminfo/g/ghostty
xterm-ghostty|ghostty|Ghostty,
        am, bce, ccc, hs, km, mc5i, mir, msgr, npc, xenl,
        colors#0x100, cols#80, it#8, lines#24, pairs#0x7fff,
        acsc=++\,\,--..00``aaffgghhiijjkkllmmnnooppqqrrssttuuvvwwxxyyzz{{||}}~~,
        bel=^G, blink=\E[5m, bold=\E[1m, cbt=\E[Z, civis=\E[?25l,
        clear=\E[H\E[2J, cnorm=\E[?12l\E[?25h, cr=\r,
        csr=\E[%i%p1%d;%p2%dr, cub=\E[%p1%dD, cub1=^H,
        cud=\E[%p1%dB, cud1=\n, cuf=\E[%p1%dC, cuf1=\E[C,
        cup=\E[%i%p1%d;%p2%dH, cuu=\E[%p1%dA, cuu1=\E[A,
        cvvis=\E[?12;25h, dch=\E[%p1%dP, dch1=\E[P, dim=\E[2m,
        dl=\E[%p1%dM, dl1=\E[M, dsl=\E]2;\007, ech=\E[%p1%dX,
        ed=\E[J, el=\E[K, el1=\E[1K, flash=\E[?5h$<100/>\E[?5l,
        fsl=^G, home=\E[H, hpa=\E[%i%p1%dG, ht=^I, hts=\EH,
        ich=\E[%p1%d@, ich1=\E[@, il=\E[%p1%dL, il1=\E[L, ind=\n,
        indn=\E[%p1%dS,
        initc=\E]4;%p1%d;rgb:%p2%{255}%*%{1000}%/%2.2X/%p3%{255}%*%{1000}%/%2.2X/%p4%{255}%*%{1000}%/%2.2X\E\\,
        invis=\E[8m, kDC=\E[3;2~, kEND=\E[1;2F, kHOM=\E[1;2H,
        kIC=\E[2;2~, kLFT=\E[1;2D, kNXT=\E[6;2~, kPRV=\E[5;2~,
        kRIT=\E[1;2C, kbs=^?, kcbt=\E[Z, kcub1=\EOD, kcud1=\EOB,
        kcuf1=\EOC, kcuu1=\EOA, kdch1=\E[3~, kend=\EOF, kent=\EOM,
        kf1=\EOP, kf10=\E[21~, kf11=\E[23~, kf12=\E[24~,
        kf13=\E[1;2P, kf14=\E[1;2Q, kf15=\E[1;2R, kf16=\E[1;2S,
        kf17=\E[15;2~, kf18=\E[17;2~, kf19=\E[18;2~, kf2=\EOQ,
        kf20=\E[19;2~, kf21=\E[20;2~, kf22=\E[21;2~,
        kf23=\E[23;2~, kf24=\E[24;2~, kf25=\E[1;5P, kf26=\E[1;5Q,
        kf27=\E[1;5R, kf28=\E[1;5S, kf29=\E[15;5~, kf3=\EOR,
        kf30=\E[17;5~, kf31=\E[18;5~, kf32=\E[19;5~,
        kf33=\E[20;5~, kf34=\E[21;5~, kf35=\E[23;5~,
        kf36=\E[24;5~, kf37=\E[1;6P, kf38=\E[1;6Q, kf39=\E[1;6R,
        kf4=\EOS, kf40=\E[1;6S, kf41=\E[15;6~, kf42=\E[17;6~,
        kf43=\E[18;6~, kf44=\E[19;6~, kf45=\E[20;6~,
        kf46=\E[21;6~, kf47=\E[23;6~, kf48=\E[24;6~,
        kf49=\E[1;3P, kf5=\E[15~, kf50=\E[1;3Q, kf51=\E[1;3R,
        kf52=\E[1;3S, kf53=\E[15;3~, kf54=\E[17;3~,
        kf55=\E[18;3~, kf56=\E[19;3~, kf57=\E[20;3~,
        kf58=\E[21;3~, kf59=\E[23;3~, kf6=\E[17~, kf60=\E[24;3~,
        kf61=\E[1;4P, kf62=\E[1;4Q, kf63=\E[1;4R, kf7=\E[18~,
        kf8=\E[19~, kf9=\E[20~, khome=\EOH, kich1=\E[2~,
        kind=\E[1;2B, kmous=\E[<, knp=\E[6~, kpp=\E[5~,
        kri=\E[1;2A, oc=\E]104\007, op=\E[39;49m, rc=\E8,
        rep=%p1%c\E[%p2%{1}%-%db, rev=\E[7m, ri=\EM,
        rin=\E[%p1%dT, ritm=\E[23m, rmacs=\E(B, rmam=\E[?7l,
        rmcup=\E[?1049l, rmir=\E[4l, rmkx=\E[?1l\E>, rmso=\E[27m,
        rmul=\E[24m, rs1=\E]\E\\\Ec, sc=\E7,
        setab=\E[%?%p1%{8}%<%t4%p1%d%e%p1%{16}%<%t10%p1%{8}%-%d%e48;5;%p1%d%;m,
        setaf=\E[%?%p1%{8}%<%t3%p1%d%e%p1%{16}%<%t9%p1%{8}%-%d%e38;5;%p1%d%;m,
        sgr=%?%p9%t\E(0%e\E(B%;\E[0%?%p6%t;1%;%?%p2%t;4%;%?%p1%p3%|%t;7%;%?%p4%t;5%;%?%p7%t;8%;m,
        sgr0=\E(B\E[m, sitm=\E[3m, smacs=\E(0, smam=\E[?7h,
        smcup=\E[?1049h, smir=\E[4h, smkx=\E[?1h\E=, smso=\E[7m,
        smul=\E[4m, tbc=\E[3g, tsl=\E]2;, u6=\E[%i%d;%dR, u7=\E[6n,
        u8=\E[?%[;0123456789]c, u9=\E[c, vpa=\E[%i%p1%dd,

**Reply by limonoct** [reply]
however the path is /home/linuxbrew/.linuxbrew/Cellar/ncurses/6.5/share/terminfo/g/ghostty,   if I look in /usr/share/terminfo/x/, it is not there,
could this be related?

**Reply by edrozenberg** [reply]
Maybe, try putting it in /usr/share. Linux brew is beyond my experience.

**Reply by dgl** [reply, marked as answer]
I suspect you've got a mix of programs using the system ncurses and the brew installed one, so it will work for tools built against the brew version, but not the system one. As @edrozenberg says putting it in the system location in `/usr/share` will work, but this is a bit confusing as you'll have to install it twice.

Better IMO would be to use a command which calls the system tic directly: `infocmp -x | ssh YOUR-SERVER -- /usr/bin/tic -x -` this will use the system version of `tic`, which will then find it can't write to `/usr/share` and install it in `~/.terminfo` which both instances of ncurses will pick up.

(This happens because the home/linuxbrew installation location is writable to your normal user in its standard install, breaking an assumption tic makes about whether it should install in the system or home directory location based on which user runs it.)

**Reply by Lira-tron** [reply]
This worked!! Thanks @dgl!!

**Comment by jpoley** [comment]
when i ssh from mac 15.2  to linux (ubuntu 22.04 LTS)  via ssh, i get double text..  when typing ls i get lsls (but command works), its just the local echo, is this something I have to set differently ?

**Reply by dgl** [reply]
@jpoley make sure you have the terminfo definition per https://ghostty.org/docs/help/terminfo#ssh
